### PR TITLE
revamp: starting the next migration 

### PR DESCRIPTION
**Pull request by soloplxya**
Stuff I have done so far: 
- Edit the `package.json` file to include dependencies and new build commands 
- Removed `react-scripts` from `package.json` file
- Installed `next` 
- Create `pages folder` 
- Create `Index.tsx` in pages folder (Moved `app.tsx` to `index.tsx`)

Problems: 
- Facing some problems with the webpack build & routing

[Link reference](https://nextjs.org/docs/migrating/from-create-react-app)

### CODE REVIEW

**Comment by Hackin7** [comment]
image

Rough functional build with Next. Modifications made:

1. Add the line `NEXT_PUBLIC_PUBLIC_URL="http://localhost:3000"` to `.env` -> Should figure out a better way to handle the 
2. `/src` directory - Has the Typescript global declaration file(needed) as well as service worker(not sure if this needed)
3. `/styles` directory, a lot of rewriting needed to be done for the CSS, but for now I combined all the CSS files into `globals.css`

This build should be functional NextJS wise

```
npm run dev
npm run build
npm run start
```

**Comment by Hackin7** [comment]
I experimented a bit with the CSS in NextJS, since there are some new bugs introduced that I'm not sure if we should try fixing yet 

Some of the current bugs involve a non-sticky nav-bar (still not sure why) - Though I think its minor

image
image

More importantly there's this layout issue, it also appears on my phone

image

There was also another issue with the cards margin (in the previous comment), but for now I hacked a fix in `_app.tsx`

I suspect the issue that resulted in these is that NextJS imports CSS differently than React (eg. Our custom css file didn't overwrite the elastic CSS)

**Comment by Hackin7** [comment]
Since NextJS and React handle CSS quite differently, and we are planning to make new pages/ redo some of the current ones, I'm wondering if we want to stick to current CSS global files for now (CSS files, and then link all together in `pages/_app.tsx`), then slowly move to either one of these as we code [\*reference](https://blog.logrocket.com/best-styling-options-nextjs/)

1. CSS module files - The method suggested by NextJS Official Documentation, but no media queries support etc.
2. Inline CSS
3. CSS code in jsx (Use `<style>` tags in the react code & insert custom CSS inside)
4. Stick with global CSS 

Some issues I have identified are

1. CSS Media Queries & Special Selectors (`:hover`), how should we handle them
2. CSS file importing - eg. our custom css files is overwriting the css code from the libraries (resulted in some bugs before)

Right now I think most of the major bugs have been fixed (can check the deploy link).
Other than CSS I think the website is functional & ready to be worked on to add features/ deployed.

File: components/Drawer.tsx [hunk 1]
```diff
@@ -116,7 +116,7 @@ const ResponsiveDrawer = (props: Props) => {
           </a>
         </Toolbar>
       </AppBar>
-      <Box
+      {/*<Box
```
Comment (wei2912) [inline components/Drawer.tsx:119]: Why was this `Box` commented out?

File: pages/_app.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+export default function App({ Component, pageProps }: AppProps) {
+  return (
+    <>
+      <style>
+        {`
+      .sui-result + .sui-result {
+        margin-top: 0px;
+      }
+      `}
```
Comment (wei2912) [inline pages/_app.tsx:16]: Think might need to run Prettier on this?

File: index.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
+    <meta name="theme-color" content="#000000" />
+    <link rel="apple-touch-icon" href="%PUBLIC_URL%/logo192.png" />
+    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+
+    <meta name="description" content="The Advisory Mentorship Program" />
+    <!-- Google -->
+    <meta itemprop="name" content="The Advisory Mentorship Program" />
+    <meta itemprop="description" content="The Advisory Mentorship Program" />
```
Comment (wei2912) [inline index.html:16]: "Program" should be "Programme". Also, could I check where the meta tags originally came from?

File: index.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    <meta
+      itemprop="image"
+      content="%PUBLIC_URL%/advisory-mentorship-logo.jpg"
+    />
+    <!-- Facebook / OpenGraph -->
+    <meta property="og:url" content="https://mentorship.advisory.sg" />
+    <meta property="og:type" content="website" />
+    <meta property="og:title" content="The Advisory Mentorship Program" />
+    <meta property="og:description" content="The Advisory Mentorship Program" />
```
Comment (wei2912) [inline index.html:25]: See above comment.

File: index.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      property="og:image"
+      content="%PUBLIC_URL%/advisory-mentorship-logo.jpg"
+    />
+    <!-- Twitter -->
+    <meta name="twitter:card" content="summary_large_image" />
+    <meta name="twitter:title" content="The Advisory Mentorship Program" />
+    <meta
+      name="twitter:description"
+      content="The Advisory Mentorship Program"
+    />
```
Comment (wei2912) [inline index.html:36]: See above comment.

File: components/Header.tsx [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import Box from "@mui/material/Box";
 
-import { advisoryMentorshipLogo, advisoryMentorshipPartners } from "../assets";
+import { advisoryMentorshipLogo, advisoryMentorshipPartners } from "./assets";
 import ResponsiveDrawer from "./Drawer";
-import "./Header.css";
+//import "./Header.css";
```
Comment (wei2912) [inline components/Header.tsx:5]: Could this line be deleted since we import all the styles under `index.ts`?

File: components/ResultView.tsx [hunk 7]
```diff
@@ -19,7 +19,7 @@ import { htmlToText } from "html-to-text";
 
 import ResultViewGrid from "./ResultViewGrid";
 import ResultViewList from "./ResultViewList";
-import "./ResultView.css";
+//import "./ResultView.css";
```
Comment (wei2912) [inline components/ResultView.tsx:22]: See above comment.

File: pages/index.tsx [hunk 8]
```diff
@@ -21,9 +21,9 @@ import ViewListIcon from "@mui/icons-material/ViewList";
 import Tabs from "@mui/material/Tabs";
 import Tab from "@mui/material/Tab";
 
-import Header from "./components/Header";
-import ResultView from "./components/ResultView";
-import "./App.css";
+import Header from "../components/Header";
+import ResultView from "../components/ResultView";
+//import "../styles/App.module.css";
```
Comment (wei2912) [inline pages/index.tsx:26]: See above line on deleting old CSS imports.

**Review by wei2912 — CHANGES_REQUESTED** [review]
Visually, the website looks the same except for a few parts:
1. Mobile navigation (this was commented out)
2. Titles look different.
3. Modal:

image
(this PR)

image
(original website)

I'm not certain why the CSS differs tbh, might need to study this more closely for `ResultViewGrid.tsx`.

File: index.html (hunk 3 above)
Comment (soloplxya) [reply index.html:16]: I think its from [here](https://github.com/AdvisorySG/mentorship-page/blob/3090e2fb7d4d1447d6b4d40d463ee9975dd77195/public/index.html)

File: pages/_app.tsx (hunk 2 above)
Comment (wei2912) [reply pages/_app.tsx:16]: Actually no, Prettier passes on this somehow.

File: index.html (hunk 3 above)
Comment (wei2912) [reply index.html]: That explains a lot, I haven't touched that file in so long that I forgot it existed

**Comment by Hackin7** [comment]
An update on the CSS
Most of the bugs was likely due to my temporary stopgap of moving CSS away from the individual component files into a global CSS file. This led to the Search UI Styles overwriting our custom CSS styles, and hence inconsistent styles

After some googling, somehow I found out a way to disable NextJS's CSS module files, such that component files can use CSS directly, just like React in the past. More information about that can be found in these links

1. https://github.com/vercel/next.js/discussions/24013
2. https://cwtuan.blogspot.com/2022/10/disable-css-module-in-nextjs-v1231-sept.html

However, this method is not really supported by the NextJS development team. Most importantly, for global CSS files, they said this

| Due to the global nature of stylesheets, and to avoid conflicts, you may only import them inside [pages/_app.js](https://nextjs.org/docs/advanced-features/custom-app).

They suggest on [their website](https://nextjs.org/docs/basic-features/built-in-css-support) to use either
1. Global CSS Files & CSS Module Files
2. CSS in JSX

Personally, I don't think this is too big of an issue, as long as we properly name our custom styles. It also brings about benefits in that we can easily modify the styles from our libraries (eg. Search UI)

File: components/Drawer.tsx (hunk 1 above)
Comment (Hackin7) [reply components/Drawer.tsx]: Should be uncommented

**Review by wei2912 — APPROVED** [review]
LGTM. As to the CSS modules issue, think we can use this method for now and slowly shift away (in line with switching to Tailwind, for consistency with github.com/AdvisorySG/dawn-advisory-theme).

**Comment by wei2912** [comment]
Merged in https://github.com/AdvisorySG/mentorship-page/commit/1adc55709baa13f3085c58711add10f2de59cd7f, closing.

**Pull request closed without merging**